src/dune_rules/coq_rules.ml

```diff
@@ -370,27 +370,31 @@ module Context = struct
     { t with boot_type }
 end
 
-let parse_coqdep ~dir ~(boot_type : Bootstrap.t) ~coq_module
-    (lines : string list) =
+let parse_coqdep ~dir ~(boot_type : Bootstrap.t) ~coq_module contents =
   let source = Coq_module.source coq_module in
-  let invalid phase =
+  let err_coqdep err =
     User_error.raise
-      [ Pp.textf "coqdep returned invalid output for %s / [phase: %s]"
+      [ Pp.textf "coqdep returned invalid output for %s"
           (Path.Build.to_string_maybe_quoted source)
-          phase
-      ; Pp.verbatim (String.concat ~sep:"\n" lines)
+      ; Pp.textf "error: %s" err
+      ; Pp.verbatim (String.concat ~sep:"\n" contents)
       ]
   in
-  let line =
-    match lines with
-    | [] | _ :: _ :: _ :: _ -> invalid "line"
-    | [ line ] -> line
-    | [ l1; _l2 ] ->
-      (* .vo is produced before .vio, this is fragile tho *)
-      l1
+  let deps =
+    List.map contents ~f:(fun line ->
+        match String.lsplit2 line ~on:':' with
+        | None -> err_coqdep "unable to parse targets"
+        | Some (targets, deps) ->
+          ( String.extract_blank_separated_words targets
+          , String.extract_blank_separated_words deps ))
   in
-  match String.lsplit2 line ~on:':' with
-  | None -> invalid "split"
+  let ext = ".vo" in
+  match
+    List.find_map deps ~f:(fun (targets, deps) ->
+        List.find_map targets ~f:(fun t ->
+            if Filename.check_suffix t ext then Some (t, deps) else None))
+  with
```

Comment:
Building the (inefficient) map just to deconstruct it is wasted work IMO, moreover this code is less resilient to changes in coqdep.

Reply:
Coqdep parsing has two modes:
- single file shot: we just need to read the first line and the first target, the rest are useless; we output a list of files
- multi-file shot: we first need to filter relevant lines, then build a (true) map with the module deps